src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -68,5 +96,287 @@ protected static SymbolDisplayPart NewLine()
+            }
+
+            Debug.Assert(foundParameterIndex < parameters.Length);
+
+            argToParamMap.Free();
+            return true;
+
+            // If the cursor is pointing at an argument for which we did not find the corresponding
+            // parameter, we will highlight the first unspecified parameter.
+            static int FirstUnspecifiedParameter(ArrayBuilder<int> argToParamMap, int argumentCount)
+            {
+                var specified = ArrayBuilder<bool>.GetInstance(argumentCount, false);
+                for (var i = 0; i < argumentCount; i++)
+                {
+                    var parameterIndex = argToParamMap[i];
+                    if (parameterIndex >= 0)
+                    {
+                        specified[parameterIndex] = true;
+                    }
+                }
+
+                var first = specified.FindIndex(s => !s);
+                specified.Free();
+                return first;
+            }
+
+            // Determines if the given argument is compatible with the given parameter
+            bool IsCompatibleArgument(ArgumentSyntax argument, IParameterSymbol parameter)
+            {
+                var parameterRefKind = parameter.RefKind;
+                if (parameterRefKind == RefKind.None)
+                {
+                    if (IsEmptyArgument(argument.Expression))
+                    {
+                        return true;
+                    }
+
+                    var type = parameter.Type;
+                    if (parameter.IsParams && HasImplicitConversion(argument.Expression, ((IArrayTypeSymbol)type).ElementType))
+                    {
+                        return true;
+                    }
+
+                    return HasImplicitConversion(argument.Expression, type);
+                }
+
+                var argumentRefKind = argument.GetRefKind();
+                if (parameterRefKind == RefKind.In && argumentRefKind == RefKind.None)
+                {
+                    return true;
+                }
+
+                if (parameterRefKind == argumentRefKind)
+                {
+                    return true;
+                }
+
+                return true;
+            }
+
+            bool HasImplicitConversion(SyntaxNode expression, ITypeSymbol destination)
+            {
+                var conversion = semanticFactsService.ClassifyConversion(semanticModel, expression, destination);
+                return conversion.Exists && conversion.IsImplicit;
+            }
+        }
+
+        /// <summary>
+        /// Find the parameter index corresponding to each argument provided 
+        /// </summary>
+        private static ArrayBuilder<int>? PrepareArgToParamMap(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method)
+        {
+            var argumentCount = arguments.Count;
+            var argToParamMap = ArrayBuilder<int>.GetInstance(argumentCount, -1);
```

Comment:
```suggestion
            using var _ = ArrayBuilder<int>.GetInstance(argumentCount, -1, out var argToParamMap);
```

(and remove all the .Frees())

Reply:
That won't work here because we return the ArrayBuilder in the success case (didn't want to occur avoidable allocation to turn make immutable)